### PR TITLE
Add an option for ButtonGroups to be unpressed

### LINKED ISSUE #6722: Allow unpressing all buttons in a ButtonGroup
### Describe the project you are working on

User interface

### Describe the problem or limitation you are having in your project

I consistently find myself wanting to have a group of buttons where only one can be pressed at a time but you can also unpress all of them. 

For instance, a map editor. When you press a button to select what object to place, you want the other buttons to unpress. Extremely easy with button groups. But what if you want to press that same button again to have no object selected? Currently this behavior is impossible without creating your own ButtonGroup-like system.

Example of map editor:

https://user-images.githubusercontent.com/66881186/233255596-2628799a-c8b0-4bfb-ba43-44140f3ad2a7.mp4

You would hit the same road block developing a UI with categories and dropdowns like this:

https://user-images.githubusercontent.com/66881186/233254988-65b61cd7-0888-4999-ae09-305e42809e22.mp4

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Solution 1: Add a toggle to ButtonGroup for whether or not all buttons can become unpressed
Solution 2: Create a child class of ButtonGroup that allows all buttons to become unpressed

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Current behavior:

https://user-images.githubusercontent.com/66881186/233247341-db02d893-0e6c-4c0f-b8a5-3c4b39fc6f29.mp4

Proposed behavior:

https://user-images.githubusercontent.com/66881186/233247374-e427a3d4-36d3-4a4e-88e4-071b99a5efbd.mp4

Possible feature UI (`allow_unpress` disabled by default to maintain compatibility):
image

PR with implementation:
https://github.com/godotengine/godot/pull/76279

### If this enhancement will not be used often, can it be worked around with a few lines of script?

It can be worked around, but the workaround is creating an entire system yourself that copies the behavior of ButtonGroup with one minor tweak. While it's not that hard, it seems unnecessary given the engine already has a system that's 99% of the way there.

### Is there a reason why this should be core and not an add-on in the asset library?

This seems like a useful enough feature to be part of the engine and it can be added without breaking compatibility. It would resolve a pain point that I'm sure many developers have run into and would make it even easier to create responsive UIs.

### PR DESCRIPTION
**Pull request by TheSecondReal0**
Closes https://github.com/godotengine/godot-proposals/issues/6722

### CODE REVIEW

File: scene/gui/base_button.cpp [hunk 1]
```diff
@@ -41,7 +41,9 @@ void BaseButton::_unpress_group() {
 	}
 
 	if (toggle_mode) {
```
Comment (KoBeWi) [inline scene/gui/base_button.cpp:43]:
```suggestion
	if (toggle_mode && !button_group->is_allow_unpress()) {
```
No need to nest this.

File: doc/classes/ButtonGroup.xml [hunk 2]
```diff
@@ -24,6 +24,9 @@
 		</method>
 	</methods>
 	<members>
+		<member name="allow_unpress" type="bool" setter="set_allow_unpress" getter="is_allow_unpress" default="false">
+			When enabled, it is possible to unpress all buttons in this [ButtonGroup]
```
Comment (KoBeWi) [inline doc/classes/ButtonGroup.xml:28]:
```suggestion
			If [code]true[/code], it is possible to unpress all buttons in this [ButtonGroup].
```

File: scene/gui/base_button.h [hunk 3]
```diff
@@ -160,6 +161,8 @@ class ButtonGroup : public Resource {
 	BaseButton *get_pressed_button();
 	void get_buttons(List<BaseButton *> *r_buttons);
 	TypedArray<BaseButton> _get_buttons();
+	void set_allow_unpress(bool value);
```
Comment (Chaosus) [inline scene/gui/base_button.h:164]:
```suggestion
	void set_allow_unpress(bool p_enabled);
```

File: scene/gui/base_button.cpp [hunk 4]
```diff
@@ -537,9 +539,20 @@ BaseButton *ButtonGroup::get_pressed_button() {
 	return nullptr;
 }
 
+void ButtonGroup::set_allow_unpress(bool value) {
+	allow_unpress = value;
```
Comment (Chaosus) [inline scene/gui/base_button.cpp:543]:
```suggestion
	allow_unpress = p_enabled;
```

File: scene/gui/base_button.cpp [hunk 5]
```diff
@@ -537,9 +539,20 @@ BaseButton *ButtonGroup::get_pressed_button() {
...
+}
+bool ButtonGroup::is_allow_unpress() {
+	return allow_unpress;
+}
+
 void ButtonGroup::_bind_methods() {
 	ClassDB::bind_method(D_METHOD("get_pressed_button"), &ButtonGroup::get_pressed_button);
 	ClassDB::bind_method(D_METHOD("get_buttons"), &ButtonGroup::_get_buttons);
+	ClassDB::bind_method(D_METHOD("set_allow_unpress", "value"), &ButtonGroup::set_allow_unpress);
```
Comment (Chaosus) [inline scene/gui/base_button.cpp:552]:
```suggestion
	ClassDB::bind_method(D_METHOD("set_allow_unpress", "enabled"), &ButtonGroup::set_allow_unpress);
```

**Review by KoBeWi — APPROVED** [review]

**Comment by kleonc** [comment]
@TheSecondReal0 A single commit is preferred, please squash the commits ([see the docs for how to rebase](https://docs.godotengine.org/en/4.0/contributing/workflow/pr_workflow.html#the-interactive-rebase)).

File: scene/gui/base_button.cpp [hunk 6]
```diff
@@ -40,8 +40,10 @@ void BaseButton::_unpress_group() {
 		return;
 	}
 
-	if (toggle_mode) {
-		status.pressed = true;
+	if (toggle_mode && !button_group->is_allow_unpress()) {
+		if (!button_group->is_allow_unpress()) {
+			status.pressed = true;
+		}
```
Comment (kleonc) [inline scene/gui/base_button.cpp:46]:
```suggestion
	if (toggle_mode && !button_group->is_allow_unpress()) {
		status.pressed = true;
```

**Review by akien-mga — APPROVED** [review]
Makes sense to me.

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**